### PR TITLE
fix(themeresource): Update VisualState.Setter value when theme changes

### PR DESCRIPTION
**Pull request by davidjohnoliver**
GitHub Issue (If applicable): fixes https://github.com/unoplatform/uno/issues/5135

## PR Type

- Bugfix

What kind of change does this PR introduce?

## PR Checklist

Please check if your PR fulfills the following requirements:

- [ ] Docs have been added/updated which fit [documentation template](https://github.com/unoplatform/uno/blob/master/doc/.feature-template.md) (for bug fixes / features)
- [x] [Unit Tests and/or UI Tests](https://github.com/unoplatform/uno/blob/master/doc/articles/uno-development/working-with-the-samples-apps.md) for the changes have been added (for bug fixes / features) (if applicable)
- [ ] Validated PR `Screenshots Compare Test Run` results.
- [x] Contains **NO** breaking changes
- [x] Associated with an issue (GitHub or internal) and uses the [automatic close keywords](https://help.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue).
- [x] Commits must be following the [Conventional Commits](https://www.conventionalcommits.org/en/v1.0.0/#summary) specification.

## Other information

Internal Issue (If applicable):

### CODE REVIEW

File: src/Uno.UI/UI/Xaml/DependencyObjectStore.Binder.cs [hunk 1]
```diff
@@ -374,8 +374,12 @@ public void SetResourceBinding(DependencyProperty dependencyProperty, object res
 			var precedence = _overriddenPrecedences?.Count > 0
 				? _overriddenPrecedences.Peek()
 				: default;
+			SetResourceBinding(dependencyProperty, resourceKey, isTheme, context, precedence, setterBindingPath: null);
+		}
 
-			var binding = new ResourceBinding(resourceKey, isTheme, context, precedence ?? DependencyPropertyValuePrecedences.Local);
+		internal void SetResourceBinding(DependencyProperty dependencyProperty, object resourceKey, bool isTheme, object context, DependencyPropertyValuePrecedences? precedence, BindingPath? setterBindingPath)
```
Comment (jeromelaban) [inline src/Uno.UI/UI/Xaml/DependencyObjectStore.Binder.cs:380]: can `resourceKey` be a string or `SpecializedResourceDictionary.ResourceKey` instead to improve resolution performance ?

File: src/Uno.UI/UI/Xaml/ResourceResolver.cs [hunk 2]
```diff
@@ -173,6 +173,34 @@ public static void ApplyResource(DependencyObject owner, DependencyProperty prop
...
+		/// </summary>
+		/// <param name="resourceKey">Key to the resource</param>
+		/// <param name="context">Optional parameter that provides parse-time context</param>
+		/// <param name="bindingPath">The binding path defined by the Setter target</param>
+		/// <param name="precedence">Value precedence</param>
+		/// <returns>
+		/// True if the value was successfully applied and registered for theme updates, false if no theme resource was found or the target
+		/// property is not a <see cref="DependencyProperty"/>.
+		/// </returns>
+		internal static bool ApplyVisualStateSetter(object resourceKey, object context, BindingPath bindingPath, DependencyPropertyValuePrecedences precedence)
```
Comment (jeromelaban) [inline src/Uno.UI/UI/Xaml/ResourceResolver.cs:187]: Same here, can we have a `ResourceKey` ?

File: src/Uno.UI/UI/Xaml/UIElement.cs [hunk 3]
```diff
@@ -822,6 +833,19 @@ internal string GetThemeSource(string propertyName)
 			}
 		}
 
+		private DependencyProperty GetDPByName(string propertyName) => GetDPByName(propertyName, this);
+
+		private static DependencyProperty GetDPByName(string propertyName, DependencyObject propertyOwner)
+		{
+			if (!propertyName.EndsWith("Property"))
```
Comment (jeromelaban) [inline src/Uno.UI/UI/Xaml/UIElement.cs:840]: Can't you use this ?

https://github.com/unoplatform/uno/blob/a610f88c0f39f894c0879bc825ec863b1cae2f55/src/Uno.UI/UI/Xaml/DependencyProperty.cs#L274

File: src/Uno.UI/UI/Xaml/UIElement.cs [hunk 4]
```diff
@@ -883,6 +907,59 @@ IEnumerable<object> TryFindResource(ResourceDictionary resourceDictionary)
...
+		///
+		/// Eg, if you call textBlock.GetPropInheritanceParent("Foreground"), it will return the first parent which has a property called
+		/// "Foreground" (most likely the Control whose template contains the TextBlock).
+		/// </summary>
+		internal FrameworkElement GetPropInheritanceParent(string propertyName)
+		{
+			var dependencyProp = GetDPByName(propertyName);
+			if (!IsInherited(dependencyProp, GetType()))
+			{
+				return null;
```
Comment (jeromelaban) [inline src/Uno.UI/UI/Xaml/UIElement.cs:922]: We should probably start moving those to a different partial class

File: src/Uno.UI/UI/Xaml/DependencyObjectStore.Binder.cs (hunk 1 above)
Comment (davidjohnoliver) [reply src/Uno.UI/UI/Xaml/DependencyObjectStore.Binder.cs:380]: Probably - I'll take a look

File: src/Uno.UI/UI/Xaml/UIElement.cs (hunk 4 above)
Comment (davidjohnoliver) [reply src/Uno.UI/UI/Xaml/UIElement.cs:922]: Makes sense, I'll move them.

Comment (davidjohnoliver) [reply src/Uno.UI/UI/Xaml/UIElement.cs]: done

File: src/Uno.UI/UI/Xaml/DependencyObjectStore.Binder.cs (hunk 1 above)
Comment (jeromelaban) [reply src/Uno.UI/UI/Xaml/DependencyObjectStore.Binder.cs]: moar performance ;)

File: src/Uno.UI/UI/Xaml/UIElement.cs (hunk 3 above)
Comment (davidjohnoliver) [reply src/Uno.UI/UI/Xaml/UIElement.cs]: fixed

File: src/Uno.UI/UI/Xaml/UIElement.debugging.cs [hunk 5]
```diff
@@ -0,0 +1,205 @@
...
+
+		/// <summary>
+		/// Debugging helper method to get a condensed summary of visual states on controls in this element's hierarchy.
+		/// </summary>
+		internal string GetVisualStatesSummary()
+		{
+			var sb = new StringBuilder();
+			foreach (var ancestor in Uno.UI.Extensions.DependencyObjectExtensions.GetAllParents(this))
+			{
+				if (ancestor is Control control && control.GetTemplateRoot() is { } root)
```
Comment (Xiaoy312) [inline src/Uno.UI/UI/Xaml/UIElement.debugging.cs:186]:
```suggestion
				if (ancestor is Control control && control.GetTemplateRoot() is IFrameworkElement root)
```

Comment (davidjohnoliver) [reply src/Uno.UI/UI/Xaml/UIElement.debugging.cs:186]: it's not necessary - `GetTemplateRoot()` is already returning `IFrameworkElement`

**Review by carldebilly — APPROVED** [review]

File: src/Uno.UI/UI/Xaml/UIElement.netstd.cs [hunk 6]
```diff
@@ -196,5 +196,55 @@ private void OnChildRemoved(UIElement child)
 
 		internal Point GetPosition(Point position, UIElement relativeTo)
 			=> TransformToVisual(relativeTo).TransformPoint(position);
+
+#if DEBUG
+
+		/// <summary>
+		/// Convenience method to find all views with the given name.
+		/// </summary>
+		public FrameworkElement[] FindViewsByName(string name) => FindViewsByName(name, searchDescendantsOnly: false);
```
Comment (ajpinedam) [inline src/Uno.UI/UI/Xaml/UIElement.netstd.cs:205]: `IEnumerable<FrameworkElement>` ?

File: src/Uno.UI/UI/Xaml/UIElement.netstd.cs [hunk 7]
```diff
@@ -196,5 +196,55 @@ private void OnChildRemoved(UIElement child)
...
+
+
+		/// <summary>
+		/// Convenience method to find all views with the given name.
+		/// </summary>
+		/// <param name="searchDescendantsOnly">If true, only look in descendants of the current view; otherwise search the entire visual tree.</param>
+		public FrameworkElement[] FindViewsByName(string name, bool searchDescendantsOnly)
```
Comment (ajpinedam) [inline src/Uno.UI/UI/Xaml/UIElement.netstd.cs:212]: `IEnumerable<FrameworkElement>` ?

**Review by ajpinedam — APPROVED** [review]

File: src/Uno.UI/UI/Xaml/UIElement.netstd.cs (hunk 6 above)
Comment (davidjohnoliver) [reply src/Uno.UI/UI/Xaml/UIElement.netstd.cs:205]: the only time you can use this is in debugging, where it's more convenient to have an array and you don't care about perf ;)

File: src/Uno.UI/UI/Xaml/SpecializedResourceDictionary.cs [hunk 8]
```diff
@@ -58,6 +58,12 @@ public ResourceKey(object key)
 					IsType = true;
 					HashCode = (uint)(t.GetHashCode() ^ IsType.GetHashCode());
 				}
+				else if (key is ResourceKey resourceKey)
+				{
+					Key = resourceKey.Key;
```
Comment (jeromelaban) [inline src/Uno.UI/UI/Xaml/SpecializedResourceDictionary.cs:63]: Throw instead, it generally hides a missing overload that uses a fast path

File: src/Uno.UI/UI/Xaml/UIElement.netstd.cs (hunk 7 above)
Comment (jeromelaban) [reply src/Uno.UI/UI/Xaml/UIElement.netstd.cs:212]: IEnumerable makes it more difficult to visualize in the debugger

**Review by jeromelaban — APPROVED** [review]

File: src/Uno.UI/UI/Xaml/UIElement.netstd.cs (hunk 6 above)
Comment (ajpinedam) [reply src/Uno.UI/UI/Xaml/UIElement.netstd.cs:205]: Agree. Debugging with `IEnumerable`s is a challenge. Thanks :)

**Pull request merged**